### ISSUE TITLE
GUI library (on canvas)

### ISSUE THREAD

**Issue by skoch9** [jupyter-widgets/pythreejs#209]
Hi, 
is there a GUI extension for pythreejs available that is similar to https://github.com/dataarts/dat.gui ?
If not, could you maybe give me some pointers how to extend pythreejs with such a library?

Thanks!

**Comment by vidartf** [comment]
I know [K3D](https://github.com/K3D-tools/K3D-jupyter) already uses this. I think this could be created as a more general widget extension:
- Point it to a widget whose traits it will control.
- Point it to a view widget (with a canvas) where it will draw the gui as an overlay.

Optional features:
- Configure a filter for which children / traits will be included in the gui.
- Configure which kinds of controls will be used for which kinds of traits.

I'd be more than happy to contribute to such a library if you start one, or I can set up the boilerplate and give you admin/contrib rights!

CC @artur-trzesiok FYI

**Comment by skoch9** [comment]
Thanks, yes I would like to implement it but first have to gain a better understanding how the python/javascript interface works. 
So what I understand is that the GUI in [K3D](https://github.com/K3D-tools/K3D-jupyter) does not allow to create/control GUI elements from the python side, is that right? The GUI controls there are coded on the javacsript side and call javascript functions. 

For our usecase we want to create the GUI elements from Python and have them call Python functions on events, similar to what ipywidgets is doing. It would help a lot to get some hints if and how this is possible from @artur-trzesiok.

**Comment by vidartf** [comment]
@skoch9 I think I can stub out a starting point with boilerplate for you, and write in how I think it could be made to work. Most of the work would need to be done on the JS side. Does that sound like an acceptable idea?

**Comment by skoch9** [comment]
@vidartf yes, that sounds like a plan. If you could create a stub, that would help me to understand the interplay and add the needed functionality. Do I understand it properly that it's technically possible to add gui elements from the python side and respond to callbacks in python?

**Comment by vidartf** [comment]
Sorry for leaving this for so long. I have now stubbed this out here: https://github.com/vidartf/jupyter-dat

**Comment by vidartf** [comment]
The relevant bits that need to be fleshed out should be discoverable by searching for "TODO" across the repo.

**Comment by skoch9** [comment]
Great thanks! I will take a look at it.
